examples/autotls/main.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"time"
+
+	"github.com/ipfs/go-log/v2"
+
+	p2pforge "github.com/ipshipyard/p2p-forge/client"
+	"github.com/libp2p/go-libp2p"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	quic "github.com/libp2p/go-libp2p/p2p/transport/quic"
+	"github.com/libp2p/go-libp2p/p2p/transport/tcp"
+	ws "github.com/libp2p/go-libp2p/p2p/transport/websocket"
+)
+
+var logger = log.Logger("autotls")
+
+func main() {
+	// Create a background context
+	ctx := context.Background()
+
+	log.SetLogLevel("*", "error")
+	log.SetLogLevel("basichost", "info")  // Set the log level for the basichost package to info
+	log.SetLogLevel("autotls", "debug")   // Set the log level for the autotls-example package to debug
+	log.SetLogLevel("p2p-forge", "debug") // Set the log level for the p2pforge package to debug
+	log.SetLogLevel("nat", "debug")       // Set the log level for the p2pforge package to debug
+
+	certLoaded := make(chan bool, 1) // Create a channel to signal when the cert is loaded
+
+	// p2pforge is the AutoTLS client library.
+	// The cert manager handles the creation and management of certificate
+	certManager, err := p2pforge.NewP2PForgeCertMgr(
+		p2pforge.WithCAEndpoint(p2pforge.DefaultCAEndpoint), // Let's Encrypt production CA. You can also use the staging CA (p2pforge.DefaultCATestEndpoint) to avoid rate limiting.
+		p2pforge.WithOnCertLoaded(func() {
+			certLoaded <- true
+		}),
+		p2pforge.WithUserAgent("go-libp2p/example/autotls"),
+	)
+
+	if err != nil {
+		panic(err)
+	}
+
+	// Start the cert manager
+	certManager.Start()
+	defer certManager.Stop()
+
+	opts := []libp2p.Option{
+		libp2p.DisableRelay(), // Disable relay, since we need a public IP address
+		libp2p.NATPortMap(),   // Attempt to open ports using UPnP for NATed hosts.
+
+		libp2p.ListenAddrStrings(
+			"/ip4/0.0.0.0/tcp/5500",         // regular tcp connections
+			"/ip4/0.0.0.0/udp/5500/quic-v1", // a UDP endpoint for the QUIC transport
+
+			// AutoTLS will automatically generate a certificate for this host
+			// and use the forge domain (`libp2p.direct`) as the SNI hostname.
+			fmt.Sprintf("/ip4/0.0.0.0/tcp/5500/tls/sni/*.%s/ws", p2pforge.DefaultForgeDomain),
+			fmt.Sprintf("/ip6/::/tcp/5500/tls/sni/*.%s/ws", p2pforge.DefaultForgeDomain),
```

Comment:
Why ipv4 + ipv6 here, but just ipv4 for the tcp listener?

Reply:
Good point. Fixed. I believe we need ipv6 TCP anyways so that AutoNAT can confirm out addresses. (We can't do port reuse for outgoing connections with WS, making it hard for us to detect our ipv6 if we don't have ipv6 TCP)